File: app/src/main/java/com/bogdan/codeforceswatcher/util/Prefs.kt
package com.bogdan.codeforceswatcher.util

import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences
import com.bogdan.codeforceswatcher.CwApp
import io.xorum.codeforceswatcher.features.auth.models.UserAccount
import io.xorum.codeforceswatcher.features.contests.models.Contest
import io.xorum.codeforceswatcher.util.Settings

class Prefs(private val context: Context) : Settings {

    override fun writeUserAccount(userAccount: UserAccount?) {
        val editor = getDefaultPrefs().edit()

        userAccount?.let {
            editor.putString(KEY_USER_ACCOUNT, it.toJson())
        } ?: editor.remove(KEY_USER_ACCOUNT)

        editor.apply()
    }

    override fun readUserAccount(): UserAccount? {
        return getDefaultPrefs().getString(KEY_USER_ACCOUNT, null)?.let {
            UserAccount.fromJson(it)
        }
    }

    override fun writeLastPinnedPostLink(pinnedPostLink: String) {
        val editor = getDefaultPrefs().edit()
        editor.putString(KEY_PINNED_POST, pinnedPostLink)
        editor.apply()
    }

    override fun readLastPinnedPostLink(): String {
        val defaultPrefs = getDefaultPrefs()
        return defaultPrefs.getString(KEY_PINNED_POST, "").orEmpty()
    }

    override fun readSpinnerSortPosition(): Int {
        val defaultPrefs = getDefaultPrefs()
        return (defaultPrefs.getString(KEY_SPINNER_SORT_POSITION, "0") ?: "0").toInt()
    }

    override fun writeSpinnerSortPosition(spinnerSortPosition: Int) {
        val editor = getDefaultPrefs().edit()
        editor.putString(KEY_SPINNER_SORT_POSITION, spinnerSortPosition.toString())
        editor.apply()
    }

    override fun readProblemsIsFavourite(): Boolean {
        val defaultPrefs = getDefaultPrefs()
        return defaultPrefs.getBoolean(KEY_PROBLEMS_IS_FAVOURITE, false)
    }

    override fun writeProblemsIsFavourite(isFavourite: Boolean) {
        val editor = getDefaultPrefs().edit()
        editor.putBoolean(KEY_PROBLEMS_IS_FAVOURITE, isFavourite)
        editor.apply()
    }

    override fun readContestsFilters(): Set<String> {
        val defaultPrefs = getDefaultPrefs()
        return defaultPrefs.getStringSet(KEY_CONTESTS_FILTERS, Contest.Platform.defaultFilterValueToSave).orEmpty()
    }

    override fun writeContestsFilters(filters: Set<String>) {
        val editor = getDefaultPrefs().edit()
        editor.putStringSet(KEY_CONTESTS_FILTERS, filters)
        editor.apply()
    }

    private fun getDefaultPrefs(): SharedPreferences {
        return context.getSharedPreferences(context.packageName, Context.MODE_PRIVATE)
    }

    override fun readProblemsTags(): List<String> {
        val defaultPrefs = getDefaultPrefs()
        return defaultPrefs.getStringSet(KEY_PROBLEMS_TAGS, setOf()).orEmpty().toList()
    }

    override fun writeProblemsTags(tags: List<String>) {
        val editor = getDefaultPrefs().edit()
        editor.putStringSet(KEY_PROBLEMS_TAGS, tags.toSet())
        editor.apply()
    }

    override fun readProblemsSelectedTags(): Set<String> {
        val defaultPrefs = getDefaultPrefs()
        return defaultPrefs.getStringSet(KEY_PROBLEMS_SELECTED_TAGS, setOf()).orEmpty()
    }

    override fun writeProblemsSelectedTags(tags: Set<String>) {
        val editor = getDefaultPrefs().edit()
        editor.putStringSet(KEY_PROBLEMS_SELECTED_TAGS, tags)
        editor.apply()
    }

    companion object {

        private const val KEY_SPINNER_SORT_POSITION = "key_counter"
        private const val KEY_PROBLEMS_IS_FAVOURITE = "key_problems_is_favourite"
        private const val KEY_CONTESTS_FILTERS = "key_contests_filters"
        private const val KEY_PINNED_POST = "key_pinned_post"
        private const val KEY_USER_ACCOUNT = "key_user_account"
        private const val KEY_PROBLEMS_TAGS = "key_problems_tags"
        private const val KEY_PROBLEMS_SELECTED_TAGS = "key_problems_selected_tags"

        @SuppressLint("StaticFieldLeak")
        private val prefs: Prefs = Prefs(CwApp.app)

        fun get() = prefs
    }
}

File: ios/algois-me/Application/utils/Prefs.swift
import Foundation
import common

fileprivate let KEY_PROBLEMS_SELECTED_TAGS = "key_problems_selected_tags"
fileprivate let KEY_PROBLEMS_TAGS = "key_problems_tags"

class Prefs: Settings {

    func readUserAccount() -> UserAccount? {
        guard let json = UserDefaults.standard.value(forKey: "userAccount") as? String else { return nil }
        
        return UserAccount.Companion().fromJson(json: json)
    }
    
    func writeUserAccount(userAccount: UserAccount?) {
        UserDefaults.standard.setValue(userAccount?.toJson(), forKey: "userAccount")
    }
  
    func readContestsFilters() -> Set<String> {
        if let savedFilters = (UserDefaults.standard.value(forKey: "contestsFilters")) as? Array<String> {
            return Set(savedFilters)
        } else {
            return Contest.PlatformCompanion().defaultFilterValueToSave
        }
    }
    
    func writeContestsFilters(filters: Set<String>) {
        UserDefaults.standard.setValue(Array(filters), forKey: "contestsFilters")
    }

    func readProblemsIsFavourite() -> Bool {
        if let savedIsFavourite = (UserDefaults.standard.value(forKey: "isFavouriteProblems")) as? Bool {
            return savedIsFavourite
        } else {
            return false
        }
    }
    
    func writeProblemsIsFavourite(isFavourite: Bool) {
        UserDefaults.standard.setValue(isFavourite, forKey: "isFavouriteProblems")
    }

    func readSpinnerSortPosition() -> Int32 {
        if let spinnerSortPosition = (UserDefaults.standard.value(forKey: "spinnerSortPosition")) as? Int32 {
            return spinnerSortPosition
        } else {
            return 0
        }
    }

    func writeSpinnerSortPosition(spinnerSortPosition: Int32) {
        UserDefaults.standard.setValue(spinnerSortPosition, forKey: "spinnerSortPosition")
    }

    func readLastPinnedPostLink() -> String {
        if let pinnedPostLink = (UserDefaults.standard.value(forKey: "pinnedPostLink")) as? String {
            return pinnedPostLink
        } else {
            return ""
        }
    }

    func writeLastPinnedPostLink(pinnedPostLink: String) {
        UserDefaults.standard.setValue(pinnedPostLink, forKey: "pinnedPostLink")
    }
    
    func resetAllDefaults() {
        let domain = Bundle.main.bundleIdentifier!
        UserDefaults.standard.removePersistentDomain(forName: domain)
        UserDefaults.standard.synchronize()
    }
    
    func readProblemsTags() -> [String] {
        UserDefaults.standard.value(forKey: KEY_PROBLEMS_TAGS) as? Array<String> ?? Array()
    }
    
    func writeProblemsTags(tags: [String]) {
        UserDefaults.standard.setValue(tags, forKey: KEY_PROBLEMS_TAGS)
    }
    
    func readProblemsSelectedTags() -> Set<String> {
        let tags = UserDefaults.standard.value(forKey: KEY_PROBLEMS_SELECTED_TAGS) as? Array<String> ?? Array()
        return Set(tags)
    }
    
    func writeProblemsSelectedTags(tags: Set<String>) {
        UserDefaults.standard.setValue(Array(tags), forKey: KEY_PROBLEMS_SELECTED_TAGS)
    }
}
